File: tools/analytics/run-monte-carlo.ts
#!/usr/bin/env node
// =============================================================================
// run-monte-carlo.ts — tool run_monte_carlo, runtime TypeScript opt-in (EP-010, US-041, TSK-073)
// =============================================================================
//
// ALTERNATIVA OPT-IN al default Python+numpy (run-monte-carlo.py). Attiva solo se
// analytics.estimation.monte_carlo.runtime: typescript. Stessa capability Task
// Analytics & Cost/Time Estimation (EP-010), stesso pattern [[thin-agents-fat-skills]]:
// è il TOOL (calcolo numerico deterministico, no LLM). Esegue il Monte Carlo
// throughput forecast (§Metodo 3) con `simple-statistics`, per factory JS-only.
//
// PATTERN.md §3 — operazione canonica opzionale «Monte Carlo Throughput».
// Wiki: wiki/concepts/task-analytics-cost-estimation-capability.md
//       [[task-analytics-cost-estimation-capability]] §Stima enterprise.
//       wiki/syntheses/task-analytics-estimation-methods.md §Metodo 3.
// ADR-026 §C — runtime TS opt-in con `simple-statistics`: cap soft 5000 iterazioni
//       (warning sopra), cap hard 50000 (fail-loud sopra), histogram a 50 bin,
//       fail-loud canonico se `simple-statistics` assente. §F `--seed` per
//       riproducibilità. Schema output equivalente al .py entro 5% di drift.
//
// INVARIANTE «mai numero puntuale»: output sempre percentili p50/p85/p95, mai mean.
// OPT-IN / R.P3: no-op a capability spenta; assenza del file non produce ERROR lint.
//
// CONTRATTO INPUT (CLI):  identico a run-monte-carlo.py
//   --throughput-samples '<JSON list>'   required
//   --backlog <int>                       required, >0
//   --iterations <int>                    default 5000 (cap soft TS, ADR-026 §C)
//   --week-count-cap <int>                default 520
//   --seed <int>                          opzionale; default random
//   --config <path>                       default "factory.config.yaml" (no-op check)
//
// CONTRATTO OUTPUT (stdout, JSON): schema equivalente al .py.
// EXIT CODES: 0 ok | 1 input/cap-hard/no-op | 2 dipendenza mancante (simple-statistics).
// =============================================================================

import * as fs from "fs";

// --- Fail-loud simple-statistics (ADR-026 §C) ------------------------------
let ss: any;
try {
  // eslint-disable-next-line @typescript-eslint/no-var-requires
  ss = require("simple-statistics");
} catch {
  process.stderr.write(
    "ERROR: simple-statistics non installato. Install:\n" +
      "  npm install --save-dev simple-statistics\n" +
      "Vedi ADR-026 §C.\n"
  );
  process.exit(2);
}

const TS_MAX_ITERATIONS_SOFT = 5000; // warning sopra
const TS_MAX_ITERATIONS_HARD = 50000; // fail-loud sopra
const HISTOGRAM_BINS = 50; // ADR-026 §C: risoluzione ridotta vs Python

function die(msg: string, code = 1): never {
  process.stderr.write(msg.replace(/\n+$/, "") + "\n");
  process.exit(code);
}

function parseArgs(argv: string[]): Record<string, string> {
  const out: Record<string, string> = {};
  for (let i = 0; i < argv.length; i++) {
    const a = argv[i];
    if (a.startsWith("--")) {
      const key = a.slice(2);
      const next = argv[i + 1];
      if (next === undefined || next.startsWith("--")) {
        out[key] = "true";
      } else {
        out[key] = next;
        i++;
      }
    }
  }
  return out;
}

// R.P3 / ADR-026 §G: no-op a capability spenta (parsing minimale tollerante).
function checkNoOp(configPath: string): void {
  if (!configPath || !fs.existsSync(configPath)) return;
  let text: string;
  try {
    text = fs.readFileSync(configPath, "utf-8");
  } catch {
    return;
  }
  let inAnalytics = false;
  let inEstimation = false;
  let enabled: boolean | null = null;
  for (const raw of text.split(/\r?\n/)) {
    const line = raw.split("#")[0].replace(/\s+$/, "");
    if (!line.trim()) continue;
    const indent = line.length - line.trimStart().length;
    const stripped = line.trim();
    if (indent === 0) {
      inAnalytics = stripped.startsWith("analytics:");
      inEstimation = false;
      continue;
    }
    if (inAnalytics && stripped.startsWith("estimation:")) {
      inEstimation = true;
      continue;
    }
    if (inAnalytics && inEstimation && stripped.startsWith("enabled:")) {
      const val = stripped.split(":")[1].trim().toLowerCase();
      enabled = ["true", "yes", "on", "1"].includes(val);
      break;
    }
    if (inEstimation && indent <= 2 && !stripped.startsWith("enabled:")) {
      inEstimation = false;
    }
  }
  if (enabled === false) {
    die(
      `Monte Carlo no-op: analytics.estimation.enabled=false in ${configPath}. ` +
        "Attivare la capability (analytics.estimation.enabled: true) o invocare " +
        "con un config diverso. Vedi ADR-026 §G / R.P3.",
      1
    );
  }
}

// PRNG deterministico (mulberry32) seedabile — ADR-026 §F.
function mulberry32(seed: number): () => number {
  let a = seed >>> 0;
  return function () {
    a |= 0;
    a = (a + 0x6d2b79f5) | 0;
    let t = Math.imul(a ^ (a >>> 15), 1 | a);
    t = (t + Math.imul(t ^ (t >>> 7), 61 | t)) ^ t;
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  };
}

function main(): void {
  const args = parseArgs(process.argv.slice(2));
  const configPath = args["config"] ?? "factory.config.yaml";
  checkNoOp(configPath);

  if (!args["throughput-samples"]) {
    die("ERROR: --throughput-samples richiesto. Es. '[1,2,3,2,4]'.");
  }
  if (!args["backlog"]) {
    die("ERROR: --backlog richiesto (intero > 0).");
  }

  let throughput: number[];
  try {
    throughput = JSON.parse(args["throughput-samples"]);
  } catch (e) {
    die(`ERROR: --throughput-samples non è JSON valido: ${e}. Atteso una lista, es. '[1,2,3,2,4]'.`);
  }
  if (!Array.isArray(throughput!) || throughput!.length === 0) {
    die("ERROR: --throughput-samples deve essere una lista non vuota di numeri.");
  }
  if (!throughput!.every((x) => typeof x === "number" && isFinite(x))) {
    die("ERROR: --throughput-samples deve contenere solo numeri.");
  }
  if (throughput!.some((x) => x < 0)) {
    die("ERROR: --throughput-samples non può contenere valori negativi.");
  }
  if (throughput!.every((x) => x === 0)) {
    die(
      "ERROR: --throughput-samples tutto a zero: throughput nullo, backlog mai " +
        "completabile. Fornire almeno un valore > 0."
    );
  }

  const backlog = parseInt(args["backlog"], 10);
  if (!Number.isInteger(backlog) || backlog <= 0) {
    die("ERROR: --backlog deve essere un intero positivo (> 0).");
  }
  let iterations = args["iterations"] ? parseInt(args["iterations"], 10) : TS_MAX_ITERATIONS_SOFT;
  if (!Number.isInteger(iterations) || iterations <= 0) {
    die("ERROR: --iterations deve essere un intero positivo (> 0).");
  }
  const weekCap = args["week-count-cap"] ? parseInt(args["week-count-cap"], 10) : 520;
  if (!Number.isInteger(weekCap) || weekCap <= 0) {
    die("ERROR: --week-count-cap deve essere un intero positivo (> 0).");
  }

  // Cap hard / soft (ADR-026 §C).
  if (iterations > TS_MAX_ITERATIONS_HARD) {
    die(
      `ERROR: Monte Carlo TS runtime non supporta N>${TS_MAX_ITERATIONS_HARD} iterations ` +
        "(richiesto: " +
        iterations +
        "); switch a runtime: python. Vedi ADR-026 §C.",
      1
    );
  }
  if (iterations > TS_MAX_ITERATIONS_SOFT) {
    process.stderr.write(
      `WARNING: --iterations ${iterations} supera il cap soft TS (${TS_MAX_ITERATIONS_SOFT}); ` +
        "performance degradate. Per N alti usare runtime: python. Vedi ADR-026 §C.\n"
    );
  }

  // Seed sempre presente (ADR-026 §F).
  const seed =
    args["seed"] !== undefined ? parseInt(args["seed"], 10) : (Math.random() * 0xffffffff) >>> 0;
  const rand = mulberry32(seed);

  const start = process.hrtime.bigint();
  const weeksNeeded: number[] = new Array(iterations);
  const n = throughput!.length;
  for (let it = 0; it < iterations; it++) {
    let completed = 0;
    let week = 0;
    while (completed < backlog && week < weekCap) {
      week++;
      completed += throughput![Math.floor(rand() * n)];
    }
    weeksNeeded[it] = week;
  }
  const durationMs = Number((process.hrtime.bigint() - start) / 1000000n);

  const p50 = Math.round(ss.quantileSorted([...weeksNeeded].sort((a, b) => a - b), 0.5));
  const sorted = [...weeksNeeded].sort((a, b) => a - b);
  const pct = (q: number) => Math.round(ss.quantileSorted(sorted, q));

  const maxWeek = Math.max(...weeksNeeded);
  const counts = new Array(maxWeek + 1).fill(0);
  for (const w of weeksNeeded) counts[w]++;
  const histogram: { week: number; count: number }[] = [];
  for (let w = 1; w <= maxWeek; w++) {
    if (counts[w] > 0) histogram.push({ week: w, count: counts[w] });
  }
  // Nota: HISTOGRAM_BINS documenta la risoluzione ridotta TS (ADR-026 §C); per
  // distribuzioni in settimane intere usiamo i bin nativi (≤ maxWeek).
  void HISTOGRAM_BINS;
  void p50;

  const result = {
    iterations,
    backlog,
    percentiles: {
      duration_weeks: { p50: pct(0.5), p85: pct(0.85), p95: pct(0.95) },
    },
    distribution: {
      histogram,
      bins: histogram.map((h) => h.week),
    },
    metadata: {
      runtime: `node-${process.versions.node}`,
      simple_statistics: (ss as any).version ?? "unknown",
      duration_ms: durationMs,
      seed,
    },
  };

  process.stdout.write(JSON.stringify(result) + "\n");
}

main();

File: tools/analytics/run-monte-carlo.py
#!/usr/bin/env python3
# =============================================================================
# run-monte-carlo.py — tool deterministico run_monte_carlo (EP-010, US-041, TSK-073)
# =============================================================================
#
# Parte della capability Task Analytics & Cost/Time Estimation (EP-010), istanza
# del pattern [[thin-agents-fat-skills-refactor]]: questo è il TOOL (logica numerica
# deterministica, no ragionamento LLM). Esegue il Monte Carlo throughput forecast
# (§Metodo 3): per `iterations` run campiona il throughput settimanale dalla
# distribuzione storica fino a coprire il backlog, raccoglie `weeks_needed` ed
# emette i percentili p50/p85/p95 della durata. Il tool NON sceglie la
# metodologia: la skill (US-040) / agente (US-043) o `force_method` decidono
# se/quando invocarlo; qui si fa solo il calcolo.
#
# PATTERN.md §3 — operazione canonica opzionale «Monte Carlo Throughput».
# Wiki: wiki/concepts/task-analytics-cost-estimation-capability.md
#       [[task-analytics-cost-estimation-capability]] §Stima enterprise (faccia
#       previsionale), §Output obbligatorio di ogni stima.
#       wiki/syntheses/task-analytics-estimation-methods.md §Metodo 3.
# ADR-026 — Monte Carlo runtime: Python + numpy come DEFAULT (performance:
#           10k simulazioni vectorizzate in <1s). §A runtime, §B fail-loud su
#           numpy/python mancante con install command esatto, §F `--seed` per
#           riproducibilità (metadata.seed sempre presente), §G backward compat.
#
# INVARIANTE «mai numero puntuale»: l'output espone SEMPRE percentili p50/p85/p95
# (mai un valore atteso singolo come primario, mai `mean`/`average`/`media`). Le
# durate sono distribuzioni a coda lunga: la media inganna.
#
# FAIL-LOUD (ADR-026 §B): se `import numpy` fallisce → exit 2 + messaggio
# canonico verbatim su stderr con `pip install` esatto. Nessun silent fallback.
#
# OPT-IN / R.P3: tool è no-op a capability spenta. La presenza/assenza del file
# non produce ERROR di lint; l'invocazione senza input obbligatori fa fail-loud.
#
# -----------------------------------------------------------------------------
# CONTRATTO INPUT (CLI)
#   --throughput-samples '<JSON list>'  distribuzione storica throughput (task/sett),
#                                       es. '[1,2,3,2,4]'. Required.
#   --backlog <int>                     numero di task da completare. Required, >0.
#   --iterations <int>                  default 10000.
#   --week-count-cap <int>              guardia anti-loop infinito, default 520 (10 anni).
#   --seed <int>                        opzionale; se assente → random da os.urandom.
#                                       Stesso seed + stesso input → output byte-identico.
#   --config <path>                     default "factory.config.yaml" (per no-op check).
#
# CONTRATTO OUTPUT (stdout, JSON puro) — schema ADR-026 §A / TSK-073:
#   { iterations, backlog,
#     percentiles: { duration_weeks: {p50, p85, p95} },
#     distribution: { histogram: [{week, count}], bins: [...] },
#     metadata: { runtime, numpy, duration_ms, seed } }
#
# EXIT CODES: 0 ok | 1 input non valido / no-op gate | 2 dipendenza mancante (numpy).
# =============================================================================

import sys
import os
import json
import time
import argparse

# --- Fail-loud numpy (ADR-026 §B) — prima riga effettiva di runtime --------
try:
    import numpy as np
except ImportError:
    sys.stderr.write(
        "ERROR: numpy non installato. Monte Carlo runtime richiede numpy >=1.24.\n"
        "\n"
        "Install:\n"
        "  pip install 'numpy>=1.24'        # globale (sconsigliato per factory)\n"
        "  # OPPURE (raccomandato — venv dedicato):\n"
        "  python3 -m venv .factory-venv/analytics\n"
        "  source .factory-venv/analytics/bin/activate\n"
        "  pip install 'numpy>=1.24'\n"
        "\n"
        "Vedi ADR-026 §B per dettaglio.\n"
    )
    sys.exit(2)


def _die(msg, code=1):
    sys.stderr.write(msg.rstrip("\n") + "\n")
    sys.exit(code)


def _check_no_op(config_path):
    """R.P3 / ADR-026 §G: no-op a capability spenta.

    Se factory.config.yaml è presente e analytics.estimation.enabled è
    esplicitamente false, il tool è no-op (non simula). Parsing minimale e
    tollerante: se il file/blocco manca o è illeggibile, NON blocchiamo
    (l'invocazione esplicita = volontà esplicita; il gate vero vive nella skill).
    """
    if not config_path or not os.path.isfile(config_path):
        return  # nessun config → non possiamo provare disabilitazione, procedi
    try:
        with open(config_path, "r", encoding="utf-8") as fh:
            text = fh.read()
    except OSError:
        return
    enabled = None
    in_analytics = in_estimation = False
    for raw in text.splitlines():
        line = raw.split("#", 1)[0].rstrip()
        if not line.strip():
            continue
        indent = len(line) - len(line.lstrip())
        stripped = line.strip()
        if indent == 0:
            in_analytics = stripped.startswith("analytics:")
            in_estimation = False
            continue
        if in_analytics and stripped.startswith("estimation:"):
            in_estimation = True
            continue
        if in_analytics and in_estimation and stripped.startswith("enabled:"):
            val = stripped.split(":", 1)[1].strip().lower()
            enabled = val in ("true", "yes", "on", "1")
            break
        # uscito dal blocco estimation se incontriamo una chiave a indent <= estimation level
        if in_estimation and indent <= 2 and not stripped.startswith("enabled:"):
            in_estimation = False
    if enabled is False:
        _die(
            "Monte Carlo no-op: analytics.estimation.enabled=false in "
            f"{config_path}. Attivare la capability (analytics.estimation.enabled: true) "
            "o invocare con un config diverso. Vedi ADR-026 §G / R.P3.",
            code=1,
        )


def simulate(throughput_samples, backlog, iterations, week_count_cap, rng):
    """Monte Carlo throughput forecast vectorizzato (synthesis §Metodo 3).

    Per ciascuna delle `iterations` run, campiona il throughput settimana per
    settimana dalla distribuzione storica fino a `tasks_completed >= backlog`,
    raccogliendo `weeks_needed`. Vectorizzato per blocchi di settimane.
    """
    samples = np.asarray(throughput_samples, dtype=np.float64)
    completed = np.zeros(iterations, dtype=np.float64)
    weeks_needed = np.zeros(iterations, dtype=np.int64)
    done = np.zeros(iterations, dtype=bool)

    week = 0
    while not done.all() and week < week_count_cap:
        week += 1
        draws = rng.choice(samples, size=iterations)
        completed = np.where(done, completed, completed + draws)
        newly_done = (~done) & (completed >= backlog)
        weeks_needed = np.where(newly_done, week, weeks_needed)
        done = done | newly_done

    # Iterazioni che non hanno raggiunto il backlog entro il cap: assegna il cap
    # (segnala throughput insufficiente; coda destra realistica, mai underestimate).
    weeks_needed = np.where(done, weeks_needed, week_count_cap)
    return weeks_needed


def main():
    parser = argparse.ArgumentParser(
        prog="run-monte-carlo.py",
        description="Monte Carlo throughput forecast (EP-010, US-041, ADR-026). "
                    "Output JSON su stdout. Invariante: sempre percentili, mai puntuale.",
    )
    parser.add_argument("--throughput-samples", required=True,
                        help="Distribuzione storica throughput come JSON list, es. '[1,2,3,2,4]'.")
    parser.add_argument("--backlog", required=True, type=int,
                        help="Numero di task da completare (>0).")
    parser.add_argument("--iterations", type=int, default=10000,
                        help="Numero di simulazioni (default 10000).")
    parser.add_argument("--week-count-cap", type=int, default=520,
                        help="Guardia anti-loop: settimane massime per run (default 520).")
    parser.add_argument("--seed", type=int, default=None,
                        help="Seed per riproducibilità; default random da os.urandom.")
    parser.add_argument("--config", default="factory.config.yaml",
                        help="Path config per no-op check (default factory.config.yaml).")
    args = parser.parse_args()

    # No-op gate (R.P3 / ADR-026 §G) prima di qualunque calcolo.
    _check_no_op(args.config)

    # --- Validazione input (fail-loud) -------------------------------------
    try:
        throughput = json.loads(args.throughput_samples)
    except json.JSONDecodeError as exc:
        _die(f"ERROR: --throughput-samples non è JSON valido: {exc}. "
             "Atteso una lista, es. '[1,2,3,2,4]'.")
    if not isinstance(throughput, list) or not throughput:
        _die("ERROR: --throughput-samples deve essere una lista non vuota di numeri, "
             "es. '[1,2,3,2,4]'.")
    try:
        throughput = [float(x) for x in throughput]
    except (TypeError, ValueError):
        _die("ERROR: --throughput-samples deve contenere solo numeri.")
    if any(x < 0 for x in throughput):
        _die("ERROR: --throughput-samples non può contenere valori negativi.")
    if all(x == 0 for x in throughput):
        _die("ERROR: --throughput-samples tutto a zero: throughput nullo, backlog mai "
             "completabile. Fornire una distribuzione storica con almeno un valore > 0.")
    if args.backlog <= 0:
        _die("ERROR: --backlog deve essere un intero positivo (> 0).")
    if args.iterations <= 0:
        _die("ERROR: --iterations deve essere un intero positivo (> 0).")
    if args.week_count_cap <= 0:
        _die("ERROR: --week-count-cap deve essere un intero positivo (> 0).")

    # --- Seed: sempre presente in metadata (ADR-026 §F) --------------------
    if args.seed is None:
        seed = int.from_bytes(os.urandom(4), "big")
    else:
        seed = int(args.seed)
    rng = np.random.default_rng(seed)

    # --- Simulazione -------------------------------------------------------
    start = time.perf_counter()
    weeks_needed = simulate(throughput, args.backlog, args.iterations,
                            args.week_count_cap, rng)
    duration_ms = int(round((time.perf_counter() - start) * 1000))

    # --- Percentili (invariante: p50/p85/p95, mai mean) --------------------
    p50, p85, p95 = (int(round(v)) for v in
                     np.percentile(weeks_needed, [50, 85, 95], method="linear"))

    # --- Histogram (settimana → conteggio) ---------------------------------
    max_week = int(weeks_needed.max())
    counts = np.bincount(weeks_needed, minlength=max_week + 1)
    histogram = [{"week": int(w), "count": int(counts[w])}
                 for w in range(1, max_week + 1) if counts[w] > 0]
    bins = [h["week"] for h in histogram]

    result = {
        "iterations": args.iterations,
        "backlog": args.backlog,
        "percentiles": {
            "duration_weeks": {"p50": p50, "p85": p85, "p95": p95}
        },
        "distribution": {
            "histogram": histogram,
            "bins": bins,
        },
        "metadata": {
            "runtime": "python-%d.%d.%d" % sys.version_info[:3],
            "numpy": np.__version__,
            "duration_ms": duration_ms,
            "seed": seed,
        },
    }

    sys.stdout.write(json.dumps(result, ensure_ascii=False) + "\n")
    return 0


if __name__ == "__main__":
    sys.exit(main())
